packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -536,25 +539,69 @@ export default class CodeGen {
+        const staticParts = this.genStaticParts(element);
+        if (staticParts) {
+            args.push(staticParts);
+        }
+
+        return this._renderApiCall(RENDER_APIS.staticFragment, args);
+    }
 
-            // has event listeners
-            if (element.listeners.length) {
-                databagProperties.push(this.genEventListeners(element.listeners));
+    genStaticParts(element: StaticElement): t.ArrayExpression | undefined {
+        const stack: (StaticElement | Text)[] = [element];
+        const partIdsToDatabagProps = new Map<number, t.Property[]>();
+        let partId = -1;
+
+        const addDatabagProp = (prop: t.Property) => {
+            let databags = partIdsToDatabagProps.get(partId);
+            if (!databags) {
+                databags = [];
+                partIdsToDatabagProps.set(partId, databags);
             }
+            databags.push(prop);
+        };
+
+        // Depth-first traversal. We assign a partId to each element, which is an integer based on traversal order.
+        while (stack.length > 0) {
+            const node = stack.shift()!;
+            partId++;
+            if (isElement(node)) {
+                // has event listeners
+                if (node.listeners.length) {
+                    addDatabagProp(this.genEventListeners(node.listeners));
+                }
 
-            // see STATIC_SAFE_DIRECTIVES for what's allowed here
-            for (const directive of element.directives) {
-                if (directive.name === 'Ref') {
-                    databagProperties.push(this.genRef(directive));
+                // see STATIC_SAFE_DIRECTIVES for what's allowed here
+                for (const directive of node.directives) {
+                    if (directive.name === 'Ref') {
+                        addDatabagProp(this.genRef(directive));
+                    }
+                }
+
+                // For depth-first traversal, prepend to the stack in reverse order
+                for (let i = node.children.length - 1; i >= 0; i--) {
+                    const childNode = node.children[i];
+                    stack.unshift(childNode);
                 }
             }
+        }
 
-            args.push(t.objectExpression(databagProperties));
+        if (partIdsToDatabagProps.size === 0) {
+            return undefined; // no databags needed
         }
 
-        return this._renderApiCall(RENDER_APIS.staticFragment, args);
+        return t.arrayExpression(
+            [...partIdsToDatabagProps.entries()].map(([partId, databagProperties]) => {
+                return this.genStaticPart(partId, databagProperties);
+            })
+        );
+    }
+
+    genStaticPart(partId: number, databagProperties: t.Property[]): t.CallExpression {
+        return this._renderApiCall(RENDER_APIS.staticPart, [
+            t.literal(partId),
+            t.objectExpression(databagProperties),
+        ]);
     }
```

Comment:
I originally considered serializing this as a function rather than as an object. This would be more similar to solid, which actually generates the minimal `firstChild`/`nextSibling` code, avoiding needing to traverse the entire tree. However, this causes several problems:

1. The `on` listeners need to be executed synchronously in order for reactivity to work properly, whereas the function would normally be called much later when the `elm` is actually ready, so we would have to do some fancy hoisting here.
2. We can't call `*Child`/`*Sibling` directly – we have to use the renderer. This makes the codegen way more complex.
3. Generating the "minimal" sibling/child traversal code is difficult for us because of the renderer, because calling `render.nextSibling(node)` prevents Terser from dead-code-elimination (unlike `node.nextSibling`). The only solution would be to mark the function as `PURE` but `astring` does not support comments in the cases where we would need it, so we would need to contribute this to `astring`.

So in general, it was way simpler to do an object here, and I'm not sure the perf gain from generating the traversal would even be worth it. It would also lead to more bloated code output versus just putting one single generic traversal function in the `engine-core`.